### PR TITLE
Feature/no gradient on discrete

### PR DESCRIPTION
**Pull request by christophernorth**
Fix for issue #91 

Simply throw an exception when you try and construct the GradientOptimiser when the graph contains latent Integer Vertices.  Simply looking for *any* Discrete Latent doesn't work as we need to be able to run the Gradient Optimiser over 'If' Vertices (and there are tests that rely on this).

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java [hunk 1]
```diff
@@ -40,6 +41,15 @@
     }
 
     public static GradientOptimizer of(BayesianNetwork bayesNet) {
+        long numIntegerVertices = bayesNet.getDiscreteLatentVertices().stream()
+            .filter(v -> v instanceof IntegerVertex)
```
Comment (rory-improbable) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]: is Integer vertex the only type of discretes we have? What about something like a categorical vertex?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java [hunk 2]
```diff
@@ -40,6 +41,15 @@
...
+            .count();
+
+        if (numIntegerVertices != 0) {
+            throw new UnsupportedOperationException("Gradient Optimization unsupported on Networks containing " +
+                "Discrete Latents");
```
Comment (rory-improbable) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]: is it worth mentioning in this error message how many offending discrete vertices there are?

Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]: I had thought about that - but given we don't really give any indication of *which* vertices are causing the issue I wasn't sure it was that useful (ie you have to look through your code to figure out what you're doing wrong anyway).  I'll add though as a bit of extra information can't hurt!  When we have Vertex labelling, we'll be able to improve this sort of error further.

Comment (rory-improbable) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]: ye good point, vertex lablels would be the only useful information maybe

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java (hunk 1 above)
Comment (rory-improbable) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]: also could extract this condition to a function called "isDiscreteVertex", which would make this more explicit, and also allow us to change the criteria for how we identify discrete vertices

Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]: If bayesNet.getDiscreteLatentVertices() returns anything other than an empty list then we should throw. Not need to filter for just integers.

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java [hunk 3]
```diff
@@ -40,6 +41,15 @@
     }
 
     public static GradientOptimizer of(BayesianNetwork bayesNet) {
+        long numIntegerVertices = bayesNet.getDiscreteLatentVertices().stream()
+            .filter(v -> v instanceof IntegerVertex)
+            .count();
+
+        if (numIntegerVertices != 0) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/GradientOptimizer.java]:
```
if(!bayesNet.getDiscreteLatentVertices().isEmpty())
```

**Pull request merged**